postgresql/resource_postgresql_schema.go

```diff
@@ -121,25 +122,57 @@ func resourcePostgreSQLSchema() *schema.Resource {
 func resourcePostgreSQLSchemaCreate(d *schema.ResourceData, meta interface{}) error {
 	c := meta.(*Client)
 
-	queries := []string{}
-
-	database := getDatabase(d, c)
-
 	c.catalogLock.Lock()
 	defer c.catalogLock.Unlock()
 
+	database := getDatabase(d, c)
 	txn, err := startTransaction(c, database)
 	if err != nil {
 		return err
 	}
 	defer deferredRollback(txn)
 
+	// If the authenticated user is not a superuser (e.g. on AWS RDS)
+	// we'll need to temporarily grant it membership in the following roles:
+	//  * the owner of the db (to have the permissions to create the schema)
+	//  * the owner of the schema, if it has one (in order to change its owner)
+	var rolesToGrant []string
+
+	dbOwner, err := getDatabaseOwner(txn, database)
+	if err != nil {
+		return err
+	}
+	rolesToGrant = append(rolesToGrant, dbOwner)
+
+	schemaOwner := d.Get("owner").(string)
+	if schemaOwner != "" && schemaOwner != dbOwner {
+		rolesToGrant = append(rolesToGrant, schemaOwner)
+
+	}
+
+	if err := withRolesGranted(txn, rolesToGrant, func() error {
+		return createSchema(d, c, txn)
+	}); err != nil {
+		return err
+	}
+
+	if err := txn.Commit(); err != nil {
+		return fmt.Errorf("Error committing schema: %w", err)
+	}
+
+	d.SetId(generateSchemaID(d, c))
+
+	return resourcePostgreSQLSchemaReadImpl(d, c)
+}
+
+func createSchema(d *schema.ResourceData, c *Client, txn *sql.Tx) error {
 	schemaName := d.Get(schemaNameAttr).(string)
 
 	// Check if previous tasks haven't already create schema
 	var foundSchema bool
-	err = txn.QueryRow(`SELECT TRUE FROM pg_catalog.pg_namespace WHERE nspname = $1`, schemaName).Scan(&foundSchema)
+	err := txn.QueryRow(`SELECT TRUE FROM pg_catalog.pg_namespace WHERE nspname = $1`, schemaName).Scan(&foundSchema)
 
+	queries := []string{}
 	if err == sql.ErrNoRows {
```

Comment:
I feel like this may leak other errors here, you don't do anything in the case that its an error but not no rows (ie network problem, etc).

Reply:
Indeed, I've just added a commit to handle correctly this error with a test to assert it works as expected.